### PR TITLE
Modular pp

### PR DESCRIPTION
**Pull request by sithhell**
This PR creates a modular HPX library that contains our utilities for preprocessor metaprogramming.

This PR includes #3582 and is a follow up to #3628 breaking the big PR into smaller commits.

### CODE REVIEW

**Comment by hkaiser** [comment]
@sithhell Thanks for separating things!

**Comment by sithhell** [comment]
FWIW, The generated docs can be found here:

 - Module documentation: https://stellar-group.gitlab.io/-/hpx/-/jobs/149308928/artifacts/build/share/hpx/docs/html/libs/pp/docs/index.html
 - Module API reference: https://stellar-group.gitlab.io/-/hpx/-/jobs/149308928/artifacts/build/share/hpx/docs/html/api.html#pp

File: libs/_example/docs/index.rst [hunk 1]
```diff
@@ -0,0 +1,54 @@
+..
+    Copyright (c) 2018 The STE||AR-Group
+
+    Distributed under the Boost Software License, Version 1.0. (See accompanying
+    file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+.. _libs__example:
+
+=========================
+Modularized HPX Libraries
```
[the remark this answers is not in the document]
Comment (sithhell) [reply libs/_example/docs/index.rst]: "HPX Modules" is probably the best term...

File: libs/pp/docs/index.rst [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+.. _libs_pp:
+
+===========
+PP
+===========
+
+This library contains useful preprocessor macros:
+
+   - :c:func:`HPX_PP_CAT`
```
[the remark this answers is not in the document]
Comment (sithhell) [reply libs/pp/docs/index.rst]: Yes, the intention was to have high level documentation there. See the docs in this PR as a start to have better documentation. What this does, essentially is to just link to the reference section. It is certainly not a good documentation ;)

**Comment by hkaiser** [comment]
@sithhell and I had a discussion today about how fine-granular these modules should be. I think these shouldn't be too fine-granular in order to avoid a maintenance nightmare in the end. I think the overall HPX set of libraries shouldn't consist out of not more than a dozen hierarchically layered modules. From this standpoint, isolating just 4 pp macros into their own module is too fine. I am however not sure how to measure this and what criteria to apply what should go into separate modules.

The other (more global) problem would be how to avoid circular dependencies between those modules (as this is the whole point of this excercise).

**Comment by msimberg** [comment]
@hkaiser, @sithhell I don't have too much of an opinion on the size of modules but if we want to avoid small modules, couldn't a good starting point be roughly one module per subdirectory in `hpx`? I count 12 of them at the moment... On the other hand, I do quite like small modules and extracting a big subdirectory into a module might end up unnecessarily challenging. A small module at a time will definitely be easier to review.

**Comment by biddisco** [comment]
Just as an aside - the VTK project (managed by Kitware, the maintainers of CMake), has recently re-implemented the module system they use. There might be some ideas we can use from the way they have setup their cmake to handle modules. One thread discussion it is here https://markmail.org/thread/cu2wyqpvporqvps6 - I have not looked at it in detail so I'm just posting this as a reminder to keep an eye on their cmake stuff as it may save us some time/effort when we upgrade ours.

**Comment by sithhell** [comment]
I agree with @msimberg in general. I think starting with as fine-grained modules as possible is the way to go. If the list starts to grow, we can still group the modules into categories etc.

For example, I am opposed to having one large util module. For that particular example, I think it is beneficial both for developing HPX itself and developing HPX applications to have a better overview of what's there and what can be reused. Factoring out headers that form a logical entity makes perfect sense here. It also simplifies to document those modules better, bit by bit.
As HPX applications are supposed to link against the "root" module/target `hpx`, we have all the freedom to reorganize the modules in the future if we hit a roadblock. Starting bit by bit, however, is IMHO the right way to go. The PP macros proposed here are a perfect example of a closed entity. Other examples would be bind (and friends) together with function (and friends) to form a `functional` module (in the style of the standard library), `test` would be another entity I'd put into a separate module, even though it would only contain one header (lightweight_test.hpp), we can bundle it with the test related cmake scripts and have a prominent section for it in the documentation (similar to what can be seen right now with the PP module).

Again, once the list of modules becomes too long, there's nothing holding us back to group the modules into categories or merge certain modules into bigger ones. However, starting with larger modules is harder, since it tends to become large PRs. In other words: Separating modules is, i think, harder than merging them

File: libs/pp/include/hpx/pp/strip_parens.hpp [hunk 3]
```diff
@@ -52,17 +48,20 @@
  *
```
Comment (hkaiser) [inline libs/pp/include/hpx/pp/strip_parens.hpp]: The link above needs to be fixed to: http://boost.2283326.n4.nabble.com/preprocessor-removing-parentheses-td2591973.html#a2591976.

Comment (sithhell) [reply libs/pp/include/hpx/pp/strip_parens.hpp]: Thanks, done!

**Comment by sithhell** [comment]
@biddisco thanks for the pointer! I think what the VTK people are doing goes far beyond what the intent of the changes proposed here is. IIUC, the VTK build system almost includes a full dependency management system. This is not what this exercise is about, as a first step, we should really try to disentangle our codebase, which is the main intention behind this effort.

File: libs/_example/docs/index.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (sithhell) [reply libs/_example/docs/index.rst]: Done.

**Comment by sithhell** [comment]
This should be ready to go now. It has been rebased onto latest master.

**Comment by hkaiser** [comment]
@sithhell we discussed creating one top-level header file for each module and use that one everywhere else. Did we actually agree to doing that?

**Comment by msimberg** [comment]
I've just updated this with latest master. Let's try to sort out the kinks and get this in soon as an example of an actual module so that we can start refining what our modules should look like.

> @sithhell we discussed creating one top-level header file for each module and use that one everywhere else. Did we actually agree to doing that?

There already is a top-level include but it's not used. Opinions on whether we want this or not?

**Comment by hkaiser** [comment]
> There already is a top-level include but it's not used. Opinions on whether we want this or not?

Including our modules through their corresponding top level header only gives us freedom for later to move things around, if needed.

**Comment by msimberg** [comment]
> Including our modules through their corresponding top level header only gives us freedom for later to move things around, if needed.

Yes. We can also decide on this later and change it without users noticing. I think the top-level headers are most useful for users anyway and they have the option of including them now.

**Comment by msimberg** [comment]
Fixed conflicts with master again. Are there any remaining concerns about this? If not I'd like to have it merged before it goes out of date compared to master again.

**Comment by hkaiser** [comment]
\o/

**Pull request merged**